FEXCore/Source/Interface/IR/Passes/x87StackOptimizationPass.cpp

```diff
@@ -0,0 +1,1148 @@
+            }
+          }
+        }
+
+        IREmit->Remove(CodeNode);
+        break;
+      }
+
+      case OP_PUSHSTACK: {
+        const auto* Op = IROp->C<IROp_PushStack>();
+        auto* SourceNode = CurrentIR.GetNode(Op->X80Src);
+
+        if (SlowPath) {
+          UpdateTopForPush_Slow();
+          StoreStackValueAtTop_Slow(SourceNode);
+        } else {
+          auto* SourceNode = CurrentIR.GetNode(Op->X80Src);
+          auto* OriginalNode = CurrentIR.GetNode(Op->OriginalValue);
+          StackData.push(StackMemberInfo {SourceNode, OriginalNode, SizeToOpSize(Op->LoadSize), Op->Float});
+        }
+        IREmit->Remove(CodeNode);
+        break;
+      }
+
+      case OP_COPYPUSHSTACK: {
+        const auto* Op = IROp->C<IROp_CopyPushStack>();
+        auto Offset = Op->StackLocation;
+        auto Value = MigrateToSlowPath_IfInvalid(Offset);
+
+        if (SlowPath) {
+          Ref St0 = LoadStackValueAtOffset_Slow(Offset);
+          UpdateTopForPush_Slow();
+          StoreStackValueAtTop_Slow(St0);
+        } else {
+          StackData.push(*Value);
+        }
+        IREmit->Remove(CodeNode);
+        break;
+      }
+
+      case OP_READSTACKVALUE: {
+        const auto* Op = IROp->C<IROp_ReadStackValue>();
+        auto Offset = Op->StackLocation;
+        auto Value = MigrateToSlowPath_IfInvalid(Offset);
+        Ref NewValue = LoadStackValue(Value, Offset);
+
+        IREmit->ReplaceUsesWithAfter(CodeNode, NewValue, CodeNode);
+        IREmit->Remove(CodeNode);
+        break;
+      }
+
+      case OP_STACKVALIDTAG: {
+        // Returns 0 if value is valid and 1 otherwise.
+        const auto* Op = IROp->C<IROp_StackValidTag>();
+        auto Offset = Op->StackLocation;
+        auto Value = MigrateToSlowPath_IfInvalid(Offset);
+
+        Ref Tag {};
+        if (SlowPath) {
+          Tag = GetX87ValidTag_Slow(Offset);
+        } else {
+          Tag = Value ? GetConstant(1) : GetConstant(0);
+        }
+
+        IREmit->ReplaceUsesWithAfter(CodeNode, Tag, CodeNode);
+        IREmit->Remove(CodeNode);
+        break;
+      }
+
+      case OP_STORESTACKMEMORY: {
+        const auto* Op = IROp->C<IROp_StoreStackMemory>();
+        auto Value = MigrateToSlowPath_IfInvalid();
+        Ref StackNode = LoadStackValue(Value);
+        Ref AddrNode = CurrentIR.GetNode(Op->Addr);
+
+        // On the fast path we can optimize memory copies.
+        // If we are doing:
+        // fld dword [rax]
+        // fst dword [rbx]
+        // We can optimize this to:
+        // ldr w2, [x0]
+        // str w2, [x1]
+        // or similar. As long as the source size and dest size are one and the same.
+        // This will avoid any conversions between source and stack element size and conversion back.
+        if (!SlowPath && Value->Source && Value->Source->first == Op->StoreSize && Value->InterpretAsFloat) {
+          LogMan::Msg::DFmt("Performing memcpy optimization from stack");
```

Comment:
Make sure this log is removed before merging since it shows up very frequently when running 32-bit games

Reply:
That's a good sign. :) I mean - that games actually use that, not necessarily that you're flooded with messages.